### PR TITLE
Add atomics for floating point types.

### LINKED ISSUE #279: `cuda::atomic_ref<float>::fetch_min` and `fetch_max` incorrect results on values of different signs
It looks like `cuda::atomic_ref<T>::fetch_min` for floating point types of `T` is not generating correct code. For these types a CAS loop implementation should be used, but as this Godbolt PTX shows, an unsigned integer `atom.max` is being generated. https://godbolt.org/z/sWeezx1a1

The problem is in the conversion to unsigned: negative and positive floats reinterpreted as unsigned integers will not compare in the same order. 

The code in the Godbolt does a max of positive values with the initial max set to a negative value. It also does a min of negative values with the initial min set to a positive value. Both result in incorrect results.

The correct output of the program should be 

```
0: Min: 0 Max: 511
1: Min: 512 Max: 1023
2: Min: 1024 Max: 1535
3: Min: 1536 Max: 2047
4: Min: 2048 Max: 2559
5: Min: 2560 Max: 3071
6: Min: 3072 Max: 3583
7: Min: 3584 Max: 4095
```

Instead, the program prints 

```
0: Min: 3.40282e+38 Max: -1
1: Min: 3.40282e+38 Max: -1
2: Min: 3.40282e+38 Max: -1
3: Min: 3.40282e+38 Max: -1
4: Min: 3.40282e+38 Max: -1
5: Min: 3.40282e+38 Max: -1
6: Min: 3.40282e+38 Max: -1
7: Min: 3.40282e+38 Max: -1
```

Tested on CUDA 11.5 on a Tesla V100 (DGX system).

Discovered when trying to convert cuSpatial to use libcudacxx `atomic_ref` rather than a custom implementation which uses atomicCAS.  RAPIDS would like to eliminate its custom atomic operation implementations, but this bug needs to be fixed first.

CC @jrhemstad

### PR DESCRIPTION
**Pull request by sleeepyjack**
This PR is a draft to add support for float/double atomics.

Please review and let me know what is missing.
Unfortunately, the diff between the old and new codegen output is a mess due to the reordering of operations.

Also rolls back #282 and fixes #279

### CODE REVIEW

File: include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_generated.h [hunk 1]
```diff
@@ -283,22 +283,170 @@ __device__ _Type __atomic_fetch_and_cuda(volatile _Type *__ptr, _Type __val, int
...
+template<class _CUDA_A, class _CUDA_B, class _CUDA_C> static inline __device__ void __cuda_fetch_add_relaxed_f32_block(_CUDA_A __ptr, _CUDA_B& __dst, _CUDA_C __op) { asm volatile("atom.add.relaxed.cta.f32 %0,[%1],%2;" : "=f"(__dst) : "l"(__ptr),"f"(__op) : "memory"); }
+template<class _CUDA_A, class _CUDA_B, class _CUDA_C> static inline __device__ void __cuda_fetch_add_release_f32_block(_CUDA_A __ptr, _CUDA_B& __dst, _CUDA_C __op) { asm volatile("atom.add.release.cta.f32 %0,[%1],%2;" : "=f"(__dst) : "l"(__ptr),"f"(__op) : "memory"); }
+template<class _CUDA_A, class _CUDA_B, class _CUDA_C> static inline __device__ void __cuda_fetch_add_volatile_f32_block(_CUDA_A __ptr, _CUDA_B& __dst, _CUDA_C __op) { asm volatile("atom.add.cta.f32 %0,[%1],%2;" : "=f"(__dst) : "l"(__ptr),"f"(__op) : "memory"); }
+template<class _Type, typename cuda::std::enable_if<sizeof(_Type)==4 and cuda::std::is_floating_point<_Type>::value, int>::type = 0>
```
Comment (wmaxey) [inline include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda_generated.h:338]: MSVC doesn't like `and`, suggest replacing with `&` despite english being possibly more appropriate to use in this case.

https://msvc.godbolt.org/z/KWWeh5zEx

File: codegen/codegen.cpp [hunk 2]
```diff
@@ -192,116 +200,130 @@ int main() {
...
+                                out << "template<class _Type, typename cuda::std::enable_if<sizeof(_Type)==" << sz/8;
+                                if(rmw.first == "exchange") {
+                                    out << ", int>::type = 0>\n";
+                                    out << "__device__ void __atomic_exchange_cuda(" << cv << "_Type *__ptr, _Type *__val, _Type *__ret, int __memorder, " << scopenametag(s.first) << ") {\n";
+                                    out << "    uint" << sz << "_t __tmp = 0;\n";
+                                    out << "    memcpy(&__tmp, __val, " << sz/8 << ");\n";
+                                }
+                                else {
+                                    if(type.first == "f")
+                                        out << " and cuda::std::is_floating_point<_Type>::value, int>::type = 0>\n";
```
Comment (wmaxey) [inline codegen/codegen.cpp:278]: Originating source of `and`.

File: codegen/codegen.cpp [hunk 3]
```diff
@@ -192,116 +200,130 @@ int main() {
...
+                                    else if(type.first == "u")
+                                        out << " and cuda::std::is_integral<_Type>::value, int>::type = 0>\n";
+                                    else
+                                        out << ", int>::type = 0>\n";
+                                    out << "__device__ _Type __atomic_" << rmw.first << "_cuda(" << cv << "_Type *__ptr, _Type __val, int __memorder, " << scopenametag(s.first) << ") {\n";
+                                    out << "    _Type __ret;\n";
+                                    if(type.first == "f" and sz == 32)
```
Comment (wmaxey) [inline codegen/codegen.cpp:285]: `and` here as well, though we don't often compile codegen.cpp on Windows.

**Comment by wmaxey** [comment]
I'm overall happy with these changes. I'll start up CI, and since Windows is not having issues I don't expect issues elsewhere either.

File: include/cuda/std/detail/libcxx/include/atomic [hunk 4]
```diff
@@ -1239,7 +1239,7 @@ _LIBCUDACXX_INLINE_VISIBILITY void __cxx_atomic_wait(__cxx_atomic_impl<_Tp, _Sco
 }
 
 // general atomic<T>/atomic_ref<T>
-template <class _Tp, int _Sco = 0, bool = is_integral<_Tp>::value && !is_same<_Tp, bool>::value>
+template <class _Tp, int _Sco = 0, bool = (is_integral<_Tp>::value || is_floating_point<_Tp>::value) && !is_same<_Tp, bool>::value>
```
Comment (wmaxey) [inline include/cuda/std/detail/libcxx/include/atomic:1242]: Because we are enabling the arithmetic operators in `cuda::std::atomics` as well we would need to extend the tests to exercise floating point.

These tests live in `tests/std/atomics/` rather than `tests/cuda/atomics`

Comment (sleeepyjack) [reply include/cuda/std/detail/libcxx/include/atomic:1242]: .upstream-tests/test/std/atomics/atomics.types.operations/atomics.types.operations.arith is this the right directory?

Comment (sleeepyjack) [reply include/cuda/std/detail/libcxx/include/atomic:1242]: Ah, I figure I need to add some more tests under the `std/atomics` directory. Basically mirror every test over integral types.

Comment (wmaxey) [reply include/cuda/std/detail/libcxx/include/atomic:1242]: Yeah, the integral tests would be a good start. Replicating the tests and doing the dispatch yourself for float/double would probably be sufficient. I don't think figuring out how to fit bitwise/arithmetic dispatches into the atomics_helpers there would be worth the effort.

**Comment by sleeepyjack** [comment]
Whups, while writing some more tests, I just stumbled over the fp min/max problem, i.e. not having specific instructions for them. I'm adjusting `codegen.cpp` so it emits CAS loop specializations.

**Comment by wmaxey** [comment]
> Whups, while writing some more tests, I just stumbled over the fp min/max problem, i.e. not having specific instructions for them. I'm gonna adjust `codegen.cpp` so it emits CAS loop specializations.

Hmm, yeah, I'll take a second pass over this component when I make fixes for the current issues as well. There's some overlap here with another problem. #279

**Comment by wmaxey** [comment]
> After extending the tests, I get some weird errors that I am yet unable to track down:
> 
```...```
> 
> @wmaxey do you know what this could be?

I am not encountering this on your latest, let me try with some other configs. Maybe some changes I had made on top of your branch fixed this.

**Comment by sleeepyjack** [comment]
> I am not encountering this on your latest, let me try with some other configs. Maybe some changes I had made on top of your branch fixed this.

For reference: I am running CUDA CTK 11.7.0 with gcc 11.2 in an ubuntu20.04 container. I only have a Pascal card available (sm_61).

**Comment by wmaxey** [comment]
We'll need to break out float add/sub into CAS loops. On MSVC we brutally cast to long*/int* which causes invalid results.

I can make these changes on top of your changes.

**Comment by wmaxey** [comment]
I've added a patch that fixes a few issues, let me know if this resolves your problems as well.

**Comment by wmaxey** [comment]
> The last four tests that still fail throw the exact same error: cudaErrorInvalidAddressSpace: operation not supported on global/shared address space.

That's a known issue for Pascal. I do not know the cause, but believe it may have something to do with an unsupported size operand. 

I think the only remaining thing would be tests similar to `atomics.types.generic/integral.pass.cpp`. We don't have CAS/ld/st coverage for floating point types. Which I'm positive works, but we should be complete. :)

**Comment by sleeepyjack** [comment]
> I think the only remaining thing would be tests similar to atomics.types.generic/integral.pass.cpp.

/done

Had to break them out into separate files although this introduces a lot of duplicate code.
We could do something like `if constexpr (std::is_integral<T>::value)` to mask out those operations in the `integral_*` tests that aren't available for fp types but this would require C++17.

**Comment by wmaxey** [comment]
Thanks for all the effort Daniel! I'll see if there's any issues with the changes again and if not I think it's okay to merge. Though I might remove some comments that documented the 'unsigned' nature of min/max.

**Comment by jrhemstad** [comment]
Would be nice to get @griwes to review this as well.

**Comment by wmaxey** [comment]
https://builds4u.nvidia.com/dvs/#/change/3154615863024800.1?eventType=Virtual&dvs_showStaging=on
https://scbuilds4u/dvs/#/change/3154615539432407.1?eventType=Virtual

DVS is clean, but SC-DVS found ICEs that I am able to repro on VC129, will try to figure out what's going on there.

**Comment by wmaxey** [comment]
```
C:\sbf\libcudacxx\.upstream-tests\test\std\atomics\atomics.types.operations\atomics.types.operations.wait\../atomics.types.operations.req/atomic_helpers.h(87): note: while compiling class template member function 'void TestEachAtomicType<TestFn,shared_memory_selector,cuda::std::__4::__detail::thread_scope_system>::operator ()(void) const'
C:\sbf\libcudacxx\.upstream-tests\test\std\atomics\atomics.types.operations\atomics.types.operations.wait\atomic_wait.pass.cpp(91): note: see reference to function template instantiation 'void TestEachAtomicType<TestFn,shared_memory_selector,cuda::std::__4::__detail::thread_scope_system>::operator ()(void) const' being compiled
C:\sbf\libcudacxx\.upstream-tests\test\std\atomics\atomics.types.operations\atomics.types.operations.wait\atomic_wait.pass.cpp(91): note: see reference to class template instantiation 'TestEachAtomicType<TestFn,shared_memory_selector,cuda::std::__4::__detail::thread_scope_system>' being compiled
INTERNAL COMPILER ERROR in 'C:\msbuild\2019\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\cl.exe'
    Please choose the Technical Support command on the Visual C++
    Help menu, or open the Technical Support help file for more information
nvcc error   : 'cl' died with status 0xC0000005 (ACCESS_VIOLATION)
# --error 0xc0000005 --
```

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_min.pass.cpp [hunk 5]
```diff
@@ -60,14 +60,59 @@ struct TestFn {
   }
 };
 
+template <template<typename, typename> typename Selector, cuda::thread_scope ThreadScope>
+struct TestFn<int, Selector, ThreadScope> {
```
Comment (griwes) [inline .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_min.pass.cpp:64]: Same as the comment above.

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_helpers.h [hunk 6]
```diff
@@ -62,6 +62,20 @@ struct TestEachIntegralType {
...
+    template<typename, typename> class Selector, cuda::thread_scope Scope
+#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 600
+    = cuda::thread_scope_system
+#endif
+>
+struct TestEachFloatingPointType {
+    __host__ __device__
+    void operator()() const {
+        TestFunctor<float, Selector, Scope>()();
+        TestFunctor<double, Selector, Scope>()();
```
Comment (griwes) [inline .upstream-tests/test/cuda/atomics/atomic.ext/atomic_helpers.h:75]: Should we also have a host-only call to `TestFunctor<long double, Selector, Scope>()()` here?

File: .upstream-tests/test/std/atomics/atomics.types.generic/floating_point.pass.cpp [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+template<typename T, cuda::thread_scope Scope>
+using cuda_std_atomic = cuda::std::atomic<T>;
+
+template<typename T, cuda::thread_scope Scope>
+using cuda_atomic = cuda::atomic<T, Scope>;
+
+int main(int, char**)
+{
+    // this test would instantiate more cases than just the ones below
```
Comment (griwes) [inline .upstream-tests/test/std/atomics/atomics.types.generic/floating_point.pass.cpp:171]: The integral tests here instantiate the test functions for all integer types. Here we only instantiate for two floating point types. It should be fine to remove this comment and have all combinations of scopes and memory selectors actually tested below.

**Review by griwes — APPROVED** [review]
In general this looks good to me; most comments I have are in tests, so I'm going to give this a :+1: and leave it up to @wmaxey as to whether these should gate landing this PR or not. The one comment I have in the actual change is rather minor.

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp [hunk 8]
```diff
@@ -60,14 +60,58 @@ struct TestFn {
...
+        assert(t.fetch_max(2) == int(1));
+        assert(t.load() == int(2));
+    }
+    // Test not greater
+    {
+        typedef cuda::atomic<int> A;
+        Selector<A, constructor_initializer> sel;
+        A & t = *sel.construct();
+        t = int(-1);
+        assert(t.fetch_max(4) == int(-1));
```
Comment (miscco) [inline .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp:90]: This does not connect with the comment. 

We are still testing a smaller `int` versus a larger threshold. Why is this changed and could we update the comment? 

If this does some horrible conversion to unsigned magic we should test that explicitly and keep the basic 3 vs 2 test

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_min.pass.cpp [hunk 9]
```diff
@@ -60,14 +60,59 @@ struct TestFn {
...
+        t = int(5);
+        assert(t.fetch_min(4) == int(5));
+        assert(t.load() == int(4));
+    }
+    {
+        typedef cuda::atomic<int> A;
+        Selector<volatile A, constructor_initializer> sel;
+        volatile A & t = *sel.construct();
+        t = int(5);
+        assert(t.fetch_min(-1) == int(5));
```
Comment (miscco) [inline .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_min.pass.cpp:81]: This is changing the values of the test, when we actually want to only add volatile. I would either keep them the same of change them consistently throughout the file.

File: .upstream-tests/test/std/atomics/atomics.types.generic/floating_point.pass.cpp [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+template <class A, class T, template<typename, typename> class Selector>
+__host__ __device__ __noinline__
+void
+do_test()
+{
+    Selector<A, constructor_initializer> sel;
+    A & obj = *sel.construct(T(0));
+    bool b0 = obj.is_lock_free();
+    ((void)b0); // mark as unused
```
[the remark this answers is not in the document]
Comment (miscco) [reply .upstream-tests/test/std/atomics/atomics.types.generic/floating_point.pass.cpp:94]: I guess the issue is that we have different answers here depending on the type. At least I would expect that `long double` would not be lock free

File: .upstream-tests/test/std/atomics/atomics.types.generic/floating_point.pass.cpp [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    obj.store(T(0));
+    assert(obj == T(0));
+    obj.store(T(1), cuda::std::memory_order_release);
+    assert(obj == T(1));
+    assert(obj.load() == T(1));
+    assert(obj.load(cuda::std::memory_order_acquire) == T(1));
+    assert(obj.exchange(T(2)) == T(1));
```
Comment (miscco) [inline .upstream-tests/test/std/atomics/atomics.types.generic/floating_point.pass.cpp:101]: I know this is the libcxx style, but could we at least add some newlines between the different functionality that is tested.

Reading a gazillion consecutive lines drains a lot of brain power

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp [hunk 12]
```diff
@@ -60,14 +60,58 @@ struct TestFn {
   }
 };
 
+template <template<typename, typename> typename Selector, cuda::thread_scope ThreadScope>
+struct TestFn<int, Selector, ThreadScope> {
```
[the remark this answers is not in the document]
Comment (wmaxey) [reply .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp:64]: Specifically, I'm attempting to get a guarantee that signed math is working as expected. It would be completely fair to split it into unsigned and signed specializations. Perhaps more tests for this API are needed. ;)

File: include/cuda/std/detail/libcxx/include/atomic [hunk 13]
```diff
@@ -2712,6 +2712,9 @@ typedef atomic<char16_t>           atomic_char16_t;
 typedef atomic<char32_t>           atomic_char32_t;
 typedef atomic<wchar_t>            atomic_wchar_t;
 
+typedef atomic<float>  atomic_float;
+typedef atomic<double> atomic_double;
```
[the remark this answers is not in the document]
Comment (wmaxey) [reply include/cuda/std/detail/libcxx/include/atomic:2721]: I'm going to say we probably don't want these. We'll need to open the door eventually for a half and just diverge further with what's available on H/D.

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp (hunk 8 above)
Comment (wmaxey) [reply .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_max.pass.cpp:90]: There's no horrible conversion, it's just specifically testing int types. The cast is, in truth, unnecessary.

File: .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_min.pass.cpp (hunk 9 above)
Comment (wmaxey) [reply .upstream-tests/test/cuda/atomics/atomic.ext/atomic_fetch_min.pass.cpp:81]: We did want to change the test in this case, but there is greater value in making a proper unsigned/signed split test so that we can guarantee that add/sub/max/min are behaving correctly.

File: include/cuda/std/detail/libcxx/include/atomic [hunk 14]
```diff
@@ -1047,10 +1047,15 @@ template<typename _Tp> struct __cxx_is_always_lock_free {
 
 #endif // defined(_LIBCUDACXX_ATOMIC_ALWAYS_LOCK_FREE)
 
-template <typename _Tp, int _Sco,
-          typename _Base = typename conditional<__cxx_is_always_lock_free<_Tp>::__value,
+template <typename _Tp, int _Sco>
+struct __cxx_atomic_impl_conditional {
```
Comment (wmaxey) [inline include/cuda/std/detail/libcxx/include/atomic:1051]: Something about `using` is cursed. This is very reminiscent of several `tuple` fixes.

Comment (wmaxey) [reply include/cuda/std/detail/libcxx/include/atomic:1051]: DVS results will be posted soon, but this fixed builds on all the compilers I was able to get repros on.

**Comment by wmaxey** [comment]
@griwes I've made the tests more straightforward, all signed types now just get extra testing and there isn't some strange int only overload.

@sleeepyjack I'd like to see the `atomic_float/double` removed. That just opens up questions about `long double` I guess.

**Comment by wmaxey** [comment]
Latest changes made sure that bitwise types are a superset of arithmetic types. I did some work to refactor the ref/non-ref classes as well, they should be easier to maintain in the future.

File: include/cuda/std/detail/libcxx/include/atomic [hunk 15]
```diff
@@ -1243,269 +1243,282 @@ _LIBCUDACXX_INLINE_VISIBILITY void __cxx_atomic_wait(__cxx_atomic_impl<_Tp, _Sco
...
-    __atomic_base& operator=(__atomic_base&&) = delete;
+    __atomic_base_storage& operator=(const __atomic_base_storage&) = default;
+    __atomic_base_storage& operator=(__atomic_base_storage&&) = default;
 
     _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR
-    __atomic_base(const _Tp& __a) _NOEXCEPT : __a_(__a) {}
+    __atomic_base_storage(_Storage&& __a) _NOEXCEPT : __a_(forward<_Storage>(__a)) {}
+};
+
+template <class _Tp, bool _Cq, typename _Storage>
```
Comment (miscco) [inline include/cuda/std/detail/libcxx/include/atomic:1261]: Could we get a more descriptive name instead of `_Cq`?

File: include/cuda/std/detail/libcxx/include/atomic [hunk 16]
```diff
@@ -1243,269 +1243,282 @@ _LIBCUDACXX_INLINE_VISIBILITY void __cxx_atomic_wait(__cxx_atomic_impl<_Tp, _Sco
...
-    __atomic_base_ref(__atomic_base_ref&&) = default;
-
-    __atomic_base_ref& operator=(const __atomic_base_ref&) = delete;
-    __atomic_base_ref& operator=(__atomic_base_ref&&) = delete;
+    __atomic_base_core& operator=(const __atomic_base_core&) = default;
+    __atomic_base_core& operator=(__atomic_base_core&&) = default;
 
     _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR
-    __atomic_base_ref(_Tp& __a) _NOEXCEPT : __a_(__a) {}
+    __atomic_base_core(_Storage&& __a) _NOEXCEPT : __atomic_base_storage<_Tp, _Storage>(forward<_Storage>(__a)) {}
```
Comment (miscco) [inline include/cuda/std/detail/libcxx/include/atomic:1396]: technically we are required to qualify `forward` as it is non ugly function

File: include/cuda/std/detail/libcxx/include/atomic [hunk 17]
```diff
@@ -1243,269 +1243,282 @@ _LIBCUDACXX_INLINE_VISIBILITY void __cxx_atomic_wait(__cxx_atomic_impl<_Tp, _Sco
...
-    mutable __cxx_atomic_impl<_Tp, _Sco> __a_;
+template <class _Tp, typename _Storage>
+struct __atomic_base_storage {
+    mutable _Storage __a_;
 
-    __atomic_base() = default;
-    __atomic_base(const __atomic_base&) = delete;
-    __atomic_base(__atomic_base&&) = delete;
+    __atomic_base_storage() = default;
+    __atomic_base_storage(const __atomic_base_storage&) = default;
```
Comment (miscco) [inline include/cuda/std/detail/libcxx/include/atomic:1251]: I am having some troubles correctly parsing the difference between the various classes. AFAICT the only difference is whether the special member functions are deleted / defaulted.

Given that the implementation of the classes is considerable, would it make sense to just derive from a single base class that does this for us, like optional does

File: include/cuda/std/detail/libcxx/include/atomic (hunk 16 above)
Comment (wmaxey) [reply include/cuda/std/detail/libcxx/include/atomic:1396]: Good catch.

File: include/cuda/std/detail/libcxx/include/atomic (hunk 15 above)
Comment (wmaxey) [reply include/cuda/std/detail/libcxx/include/atomic:1261]: I'll push an update with `_Cq`->`_ConstQualified`

**Review by griwes — APPROVED** [review]

**Comment by sleeepyjack** [comment]
Wooohoo, first contribution merged ☑️
